src/librustc/mir/cache.rs

```diff
@@ -1,71 +1,261 @@
+            for (bb, data) in body.basic_blocks().iter_enumerated() {
+                if let Some(ref term) = data.terminator {
+                    for &tgt in term.successors() {
+                        result[tgt].push(bb);
+                    }
+                }
+            }
+
+            self.predecessors = Some(result)
+        }
+
+        self.predecessors.as_ref().unwrap()
+    }
+
+    #[inline]
+    pub fn predecessors_for(&mut self, bb: BasicBlock, body: &Body<'_>) -> &[BasicBlock] {
+        &self.predecessors(body)[bb]
+    }
+
+    #[inline]
+    pub fn predecessor_locations<'a>(&'a mut self, loc: Location, body: &'a Body<'a>) -> impl Iterator<Item = Location> + 'a {
+        let if_zero_locations = if loc.statement_index == 0 {
+            let predecessor_blocks = self.predecessors_for(loc.block, body);
+            let num_predecessor_blocks = predecessor_blocks.len();
+            Some(
+                (0..num_predecessor_blocks)
+                    .map(move |i| predecessor_blocks[i])
+                    .map(move |bb| body.terminator_loc(bb)),
+            )
+        } else {
+            None
+        };
+
+        let if_not_zero_locations = if loc.statement_index == 0 {
+            None
+        } else {
+            Some(Location { block: loc.block, statement_index: loc.statement_index - 1 })
+        };
+
+        if_zero_locations.into_iter().flatten().chain(if_not_zero_locations)
+    }
+
+    #[inline]
+    pub fn basic_blocks_mut<'a, 'tcx>(&mut self, body: &'a mut Body<'tcx>) -> &'a mut IndexVec<BasicBlock, BasicBlockData<'tcx>> {
+        debug!("bbm: Clearing predecessors cache for body at: {:?}", body.span.data());
+        self.invalidate_predecessors();
+        &mut body.basic_blocks
+    }
+
+    #[inline]
+    pub fn basic_blocks_and_local_decls_mut<'a, 'tcx>(
+        &mut self,
+        body: &'a mut Body<'tcx>
+    ) -> (&'a mut IndexVec<BasicBlock, BasicBlockData<'tcx>>, &'a mut LocalDecls<'tcx>) {
+        debug!("bbaldm: Clearing predecessors cache for body at: {:?}", body.span.data());
+        self.invalidate_predecessors();
+        (&mut body.basic_blocks, &mut body.local_decls)
     }
 }
 
-impl rustc_serialize::Decodable for Cache {
-    fn decode<D: Decoder>(d: &mut D) -> Result<Self, D::Error> {
-        Decodable::decode(d).map(|_v: ()| Self::new())
+pub struct BodyCache<T> {
+    cache: Cache,
+    body: T,
+}
+
+impl<T> BodyCache<T> {
+    pub fn new(body: T) -> Self {
+        Self {
+            cache: Cache::new(),
+            body
+        }
     }
 }
 
-impl<'a> HashStable<StableHashingContext<'a>> for Cache {
-    fn hash_stable(&self, _: &mut StableHashingContext<'a>, _: &mut StableHasher) {
-        // Do nothing.
+impl<'a, 'tcx> BodyCache<&'a Body<'tcx>> {
+    #[inline]
+    pub fn predecessors_for(&mut self, bb: BasicBlock) -> &[BasicBlock] {
+        self.cache.predecessors_for(bb, self.body)
+    }
+
+    #[inline]
+    pub fn body(&self) -> &'a Body<'tcx> {
```

Comment:
I would have thought your lifetime troubles came from the fact that the `body` function was using the wrong lifetime (or none), but it has `'a`, which is unrelated to the anonymous lifetime on `&self`. I'm not sure why the code using the `body` method needs two distinct lifetimes

Reply:
The errors were related to the constraints around the `BuilderMethod` trait.
```rust
pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
    cx: &'a Bx::CodegenCx,
    llfn: Bx::Function,
    mut mir: BodyCache<&'a Body<'tcx>>,
    ...)
```

When the above `mir` was `mir: &mut BodyCache<&'a Body<'tcx>>`, then there were issues resolving the lifetime of `'a` and of `mir`. I don't have the exact error anymore because the last rebase I did introduced new issues I had to work around (already live in this PR).